### PR TITLE
🚨 [gemma 3/4] Fix bidirectional attention masking crossing sliding window boundaries

### PR DESCRIPTION
**Pull request by douglas-reid**
[CI](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=46850)

# What does this PR do?

This fixes the behavior of attention masking for image tokens to properly respect sliding window boundaries for Gemma 3/4 models. This was needed to match the intended model behavior for local layers.

- [X] I confirm that this is not a pure code agent PR.

## Before submitting
- [X] Did you read the [contributor guideline](https://huggingface.co/docs/transformers/contributing) and the
      [Pull Request](https://huggingface.co/docs/transformers/pr_checks) checks?
- [X] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes according to the [guidelines](https://github.com/huggingface/transformers/tree/main/docs)?
- [X] Did you write any new necessary [tests](https://huggingface.co/docs/transformers/testing)?

---

### CI recap

**Dashboard:** [View test results in Grafana](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=46850)
**Latest run:** [28405951213:2](https://github.com/huggingface/transformers/actions/runs/28406342486)
**Result:** `success` | **Jobs:** 7 | **Tests:** 2,337 | **Failures:** 0 | **Duration:** 32m 59s

### CODE REVIEW

File: src/transformers/models/gemma3/modeling_gemma3.py [hunk 1]
```diff
@@ -709,6 +715,46 @@ def get_block_sequence_ids_for_mask(token_type_ids: torch.Tensor, device: torch.
...
+
+    # Sliding attention: AND(sliding_window, OR(causal, blockwise))
+    # Pass blockwise as or_mask_function (applied as step 2 in create_causal_mask)
+    # Pass sliding_window as and_mask_function (applied as step 3, after OR)
+    # Do NOT pass block_sequence_ids (to avoid the incorrect step 4 final OR)
+    sliding_mask = create_causal_mask(
+        **mask_kwargs,
+        or_mask_function=blockwise_overlay(block_sequence_ids),
+        and_mask_function=sliding_window_overlay(config.sliding_window),
+    )
```
Comment (zucchini-nlp) [inline src/transformers/models/gemma3/modeling_gemma3.py:764]: passing or/and functions will trigger torch.vmap path, which might add some overhead. Gemma models are the only ones using `blockwise_overlay` for sliding attention

I am seeing we updated all except for paligemma, does that mean paligemma-2 needs to keep full-mask in prefix for sliding layers?

Comment (zucchini-nlp) [reply src/transformers/models/gemma3/modeling_gemma3.py:764]: I am thinking that we might change the internal `create_sliding_window_mask` if the pattern is used only by gemma models and in the same way

Comment (douglas-reid) [reply src/transformers/models/gemma3/modeling_gemma3.py:764]: i didn't think about paligemma-2. it probably needs to be updated too. How should we proceed here? Update the internal implementation, or continue with the custom fixes?

Comment (zucchini-nlp) [reply src/transformers/models/gemma3/modeling_gemma3.py:764]: Can you check for Paligemma-2 in which case we could update the internal implementation imo. If PG2 is correct in current `main` with full bidirectional mask in sliding layers, we can keep the PR as there isn't much we can do

Comment (douglas-reid) [reply src/transformers/models/gemma3/modeling_gemma3.py:764]: afaict, paligemma-2 does not need any changes.

File: tests/models/gemma3/test_modeling_gemma3.py [hunk 2]
```diff
@@ -427,6 +427,52 @@ def test_flash_attn_3_from_config(self):
     def test_flash_attn_4_from_config(self):
         self.flash_attn_from_config(attn_implementation="flash_attention_4", test_fwd_in_train=False)
 
+    def test_attention_mask_composition(self):
+        from transformers.models.gemma3.modeling_gemma3 import create_masks_for_vision_model
```
Comment (zucchini-nlp) [inline tests/models/gemma3/test_modeling_gemma3.py]: could you move import to the top of the file

**Review by zucchini-nlp — APPROVED** [review]
Oke, so we have different masks for Gemma3 and Gemma4 at the end. I think this is change the expected outputs in slow CI, let me trigger it. Some fast tests on assisted decoding are failing btw

Also lets mark the PR with 🚨 , as it is breaking even if in a good way

**Comment by zucchini-nlp** [comment]
run-slow: gemma3, gemma4, gemma4_unified

File: tests/models/gemma3/test_modeling_gemma3.py (hunk 2 above)
Comment (douglas-reid) [reply tests/models/gemma3/test_modeling_gemma3.py]: Should be done now.

**Comment by douglas-reid** [comment]
Applied some test updates and added 🚨 to the PR title.

**Comment by zucchini-nlp** [comment]
run-slow: gemma3, gemma4, gemma4_unified
